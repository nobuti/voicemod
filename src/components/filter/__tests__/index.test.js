import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { Filter } from "../";

jest.mock("../../../utils/debounce", () => jest.fn((fn) => fn));

describe("Filter", () => {
  const selectRandom = jest.fn();
  const updateFilter = jest.fn();

  const filter = {
    search: "",
    categorie: "all",
    sort: "asc",
  };

  const voices = [
    { id: 1, name: "zerous", tags: ["one"] },
    { id: 2, name: "cebolleta", tags: ["two"] },
    { id: 3, name: "wadus", tags: ["one", "two"] },
    { id: 4, name: "1x2", tags: ["three"] },
  ];

  test("should render properly", () => {
    const { container } = render(
      <Filter
        filter={filter}
        updateFilter={updateFilter}
        voices={voices}
        onRandom={selectRandom}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId("filter")).toBeInTheDocument();
  });

  test("should show filter state properly", () => {
    const { rerender, container } = render(
      <Filter
        filter={filter}
        updateFilter={updateFilter}
        voices={voices}
        onRandom={selectRandom}
      />
    );

    const updatedFilter = {
      ...filter,
      categorie: "one",
    };

    rerender(
      <Filter
        filter={updatedFilter}
        updateFilter={updateFilter}
        voices={voices}
        onRandom={selectRandom}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should update filter state properly", () => {
    render(
      <Filter
        filter={filter}
        updateFilter={updateFilter}
        voices={voices}
        onRandom={selectRandom}
      />
    );

    fireEvent.change(screen.getByTestId("search-field"), {
      target: { value: "wadus" },
    });

    expect(updateFilter).toHaveBeenCalledWith({
      filter: "search",
      value: "wadus",
    });
  });

  test("should select random voice properly", () => {
    render(
      <Filter
        filter={filter}
        updateFilter={updateFilter}
        voices={voices}
        onRandom={selectRandom}
      />
    );
    fireEvent.click(screen.queryByTestId("filter-random"));
    expect(selectRandom).toHaveBeenCalled();
  });
});
