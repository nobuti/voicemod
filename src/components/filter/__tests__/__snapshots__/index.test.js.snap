// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should render properly 1`] = `
<div>
  <header
    class="Filter"
    data-testid="filter"
  >
    <div
      class="Filter-search"
    >
      <div
        class="Search"
        data-testid="search"
      >
        <i
          class="Search-icon"
          data-testid="search-icon"
        />
        <input
          class="Search-field"
          data-testid="search-field"
          name="search"
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="Filter-selectors"
    >
      <div
        class="Filter-withIcon Filter-select"
      >
        <i
          class="Filter-icon Filter-iconTag"
        />
        <div
          class="Select"
          data-testid="select"
          tabindex="0"
        >
          <div
            class="Select-inner"
          >
            <button
              class="Select-value"
              data-testid="select-value"
            >
              All
            </button>
            <i
              class="Select-icon"
              data-testid="select-icon"
            />
          </div>
        </div>
      </div>
      <div
        class="Filter-withIcon Filter-select"
      >
        <i
          class="Filter-icon Filter-iconSort"
        />
        <div
          class="Select"
          data-testid="select"
          tabindex="0"
        >
          <div
            class="Select-inner"
          >
            <button
              class="Select-value"
              data-testid="select-value"
            >
              Ascending
            </button>
            <i
              class="Select-icon"
              data-testid="select-icon"
            />
          </div>
        </div>
      </div>
      <button
        class="Filter-button Filter-withIcon"
        data-testid="filter-random"
      >
        <i
          class="Filter-icon Filter-iconRandom"
        />
        <span>
          Select a random voice
        </span>
      </button>
    </div>
  </header>
</div>
`;

exports[`Filter should show filter state properly 1`] = `
<div>
  <header
    class="Filter"
    data-testid="filter"
  >
    <div
      class="Filter-search"
    >
      <div
        class="Search"
        data-testid="search"
      >
        <i
          class="Search-icon"
          data-testid="search-icon"
        />
        <input
          class="Search-field"
          data-testid="search-field"
          name="search"
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="Filter-selectors"
    >
      <div
        class="Filter-withIcon Filter-select"
      >
        <i
          class="Filter-icon Filter-iconTag"
        />
        <div
          class="Select"
          data-testid="select"
          tabindex="0"
        >
          <div
            class="Select-inner"
          >
            <button
              class="Select-value"
              data-testid="select-value"
            >
              one
            </button>
            <i
              class="Select-icon"
              data-testid="select-icon"
            />
          </div>
        </div>
      </div>
      <div
        class="Filter-withIcon Filter-select"
      >
        <i
          class="Filter-icon Filter-iconSort"
        />
        <div
          class="Select"
          data-testid="select"
          tabindex="0"
        >
          <div
            class="Select-inner"
          >
            <button
              class="Select-value"
              data-testid="select-value"
            >
              Ascending
            </button>
            <i
              class="Select-icon"
              data-testid="select-icon"
            />
          </div>
        </div>
      </div>
      <button
        class="Filter-button Filter-withIcon"
        data-testid="filter-random"
      >
        <i
          class="Filter-icon Filter-iconRandom"
        />
        <span>
          Select a random voice
        </span>
      </button>
    </div>
  </header>
</div>
`;
